New Major Release (v3.0.0)

delofson0211 opened the pull request:
- Changing to threadsafe code (compatible with Geant4.10.7)
- incorporating Global Singleton ConfigManager
- break AnalysisManager into 2 pieces: AnalysisManager and AnalysisData
- Incorporate GENIEManager to input Genie files
- Include a flag to toggle EventTimeWindow Cut
- Include messenger in ConfigManager to allow for toggle between VD and HD Detector Geometry
- Change variable type of number_hits_ from Int to Long64_t

delofson0211 committed:
adding getopt to G4_QPIX.cpp

delofson0211 committed:
Reverted app/G4_QPIX.cpp back to geant4 friendly format

delofson0211 committed:
Added new files ROOTManager, ConfigManager, ActionInitialization

delofson0211 committed:
Update src/CMakeLists.txt to include new files

delofson0211 committed:
Split AnalysisManager into AnalysisManager and AnalysisData. Made other src files compatible with ConfigManager

delofson0211 committed:
Updated src/CMakeLists.txt to include AnalysisData.cpp

delofson0211 committed:
removed .swo file

delofson0211 committed:
added forward declaration of AnalysisData in the .h file

delofson0211 committed:
Debugged to allow to compile

delofson0211 committed:
Created new Template_GENIE.mac

delofson0211 committed:
removed default marley_json

delofson0211 committed:
First working version of qpixg4 with GENIE input file. Needs lot of fine tuning and clean up

delofson0211 committed:
Merge pull request #9 from Q-Pix/master

Keeping Develop up to date with master

delofson0211 committed:
Merge branch 'Q-Pix:master' into feature/3

delofson0211 committed:
Cleaned up print outs, renamed ROOTManager->GENIEManager

delofson0211 committed:
Fixed GENIEManager indices. Producing reasonable output. Added the intermediate generator particles

delofson0211 committed:
Fixed some typos

delofson0211 committed:
qpixg4 produces output with genie files. commented out config print commands

delofson0211 committed:
Changed GENIEManager to thread local singleton

delofson0211 committed:
dropped unnecessary mutexes now that GENIEManager is a thread local singleton

delofson0211 committed:
updated genie macro, fixed some issues

delofson0211 committed:
Merge pull request #10 from delofson0211/feature/3

New qpixg4 with genie capabilities

delofson0211 committed:
Bandaid on User Limit issue from step->GetPostStepPoint()->GetProcessDefinedStep() null pointer

delofson0211 committed:
Merge pull request #11 from delofson0211/feature/3

Bandaid on User Limit issue from step->GetPostStepPoint()->GetProcess…

delofson0211 committed:
randomizing seeds. Some fixed typos

delofson0211 committed:
Merge pull request #12 from delofson0211/feature/3

randomizing seeds. Some fixed typos

delofson0211 committed:
fixed the generatorParticle/RootParticle momentum issue

delofson0211 committed:
Merge pull request #13 from delofson0211/feature/3

Adding GENIEManager

Commit:
Change to allow VD geometry

Commit:
Change single_muon.mac to have HD configuration as a default option

delofson0211 committed:
Merge pull request #16 from Q-Pix/feature/VD_configuration

Change to allow VD geometry

delofson0211 committed:
adding ConfigManager, allowing for GENIE file input, making thread safe, and changing number_hits_ from Int to Long64_t to allow for larger events, incorporating addition with HD/VD detector construction

asanchezcastillo commented on src/AnalysisManager.cpp:
```diff
@@ -27,320 +48,159 @@ AnalysisManager::~AnalysisManager()
+  detector_length_x_ = detector_length_x;
+  detector_length_y_ = detector_length_y;
+  detector_length_z_ = detector_length_z;
+  detectorConfiguration_ = ConfigManager::GetDetectorConfiguration();
```
This method is not implemented in the ConfigManager class.

kkeefe replied:
looks like this is resolved with `ConfigManager::Instance();`
```suggestion
  detectorConfiguration_ = ConfigManager::Instance();
```

delofson0211 replied:
There should be more in ConfigManager.h. Let me see whats going on with this

kelseymh replied:
These changes were reverted and fixed properly by adding some uncommitted changes from Dave's clone.

asanchezcastillo commented on src/AnalysisManager.cpp:
```diff
@@ -27,320 +48,159 @@ AnalysisManager::~AnalysisManager()
 
 //-----------------------------------------------------------------------------
-void AnalysisManager::FillMetadata(double const & detector_length_x,
+void AnalysisManager::FillMetadata(bool const & detectorConfiguration, double const & detector_length_x,
```
The detectorConfiguration argument is no longer needed if we use the ConfigManager.

asanchezcastillo commented on src/DetectorConstruction.h:
```diff
@@ -19,9 +19,13 @@ class DetectorConstruction: public G4VUserDetectorConstruction
 public:
   DetectorConstruction();
   virtual ~DetectorConstruction();
+  bool GetDetectorConfiguration() const { return detectorConfiguration;}
```
I assume this method is no longer needed if we use the ConfigManager.

kkeefe replied:
It looks like this is slightly different than what the ConfigManager is doing.. It looks like `detectorConfiguration` within `DetectorConstruction.h` is what is used to select between VD and HD detector  constructions. It also appears that there's no way to set this to false, and thus allowing for a VD construction.

This method more or less returns true if HD was constructed, and false if VD. But I don't think there's a way to set `detectorConfiguration` False.

kkeefe replied:
Included a commit which modifies the constructor to allow selection between VD and HD.

asanchezcastillo reviewed (commented):
I could not compile the code as the method GetDetectorConfiguration is not present in the ConfigManager class.

kkeefe committed:
Update src/AnalysisManager.cpp

ConfigManager uses instance method tell AnalysisManager if it exists

kkeefe commented on src/DetectorConstruction.cpp:
```diff
@@ -18,18 +18,22 @@
-
-DetectorConstruction::DetectorConstruction(): G4VUserDetectorConstruction()
-{}
+DetectorConstruction::DetectorConstruction(): G4VUserDetectorConstruction(), detectorConfiguration(true)
```
this should be parameterized.

kkeefe commented on src/DetectorConstruction.cpp:
```diff
@@ -18,18 +18,22 @@
-
-DetectorConstruction::DetectorConstruction(): G4VUserDetectorConstruction()
-{}
+DetectorConstruction::DetectorConstruction(): G4VUserDetectorConstruction(), detectorConfiguration(true)
+{
+  fMessenger = new G4GenericMessenger(this, "/Geometry/", "Geometry configuration");
+  fMessenger->DeclareProperty("DetectorConfiguration", detectorConfiguration, "True if HD, false if VD");
+}
```
```suggestion
DetectorConstruction::DetectorConstruction(const bool& useHorizontalDrift/*=True*/): G4VUserDetectorConstruction(), detectorConfiguration(useHorizontalDrift)
{
  fMessenger = new G4GenericMessenger(this, "/Geometry/", "Geometry configuration");
  fMessenger->DeclareProperty("DetectorConfiguration", detectorConfiguration, "True if HD, false if VD");
}
```

asanchezcastillo replied:
The value of detectorConfiguration is initialized to true by default and afterwards read from the .mac file. I don't see why we should pass `const bool& useHorizontalDrift=True` as an argument.

kkeefe commented on src/DetectorConstruction.h:
```diff
@@ -19,9 +19,13 @@ class DetectorConstruction: public G4VUserDetectorConstruction
 public:
   DetectorConstruction();
```
```suggestion
  DetectorConstruction(const bool& useHorizontalDrift=True);
 ```

kkeefe committed:
Apply suggestions from code review

allow detector construction to use a bool argument to select between HD and VD constructions

delofson0211 commented:
@kkeefe @asanchezcastillo Thank you both for checking this. I have more changes on my local that were supposed to get pushed and apparently never did. Let me get the rest of that up and hopefully it will seem more logical

delofson0211 committed:
undo changes introduced during review

delofson0211 committed:
Adding changes to allow for HD/VD detector toggling

delofson0211 committed:
changed setup_marley.sh to use preset /scratch/user/dave.elofson/software/marley variable or user input variable

delofson0211 commented:
Try running again. things should work this time

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
+
 
+#include "G4RunManagerFactory.hh"
 #include <G4RunManager.hh>
```
If you're making changes, it'd be good to put _all_ the Geant4, ROOT, etc. `#include`s into quotation marks, not angle brackets.  Formally, the angle brackets are only for system-level includes (e.g., `#include <stdlib.h>` or `#include <iostream>`.  Most compilers nowadays ignore that distinction.  But consistency can help others with understanding.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
+#include <G4OpticalPhysics.hh>
 
+#include "TROOT.h"
 
```
I'd get rid of spurious blank lines where possible.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
 int main(int argc, char** argv)
 {
 
+  ConfigManager * configManager = ConfigManager::Instance();
```
If this is being used strictly to ensure that the ConfigManager gets instantiated at the start of the job, you don't need to assign the call to a variable (right now, you're getting a compiler warning about "unused variable", which you can eliminate).

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
 {
 
+  ConfigManager * configManager = ConfigManager::Instance();
+  ConfigManager::Print();
```
Do you really want to dump the default configuration before the user has had a chance to set any of the parameters?  That's nice for debugging, but I imagine that in production you want to limit it to users who put a "printConfig" (or whatever) macro command into their setup.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
   CLHEP::HepRandom::setTheEngine(new CLHEP::RanecuEngine());
   //set random seed with system time
   G4long seed = time(NULL);
-  CLHEP::HepRandom::setTheSeed(seed);
```
I would put this back for production, but we should talk offline about how to do this better when you are in a production environment with multiple jobs running in parallel.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
   //set random seed with system time
   G4long seed = time(NULL);
-  CLHEP::HepRandom::setTheSeed(seed);
+  //CLHEP::HepRandom::setTheSeed(seed);
```
I recommend not leaving commented out lines when you have a Git repository.  You can revert to an earlier version by checking out or cherry-picking from a past commit.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
+  ConfigManager * configManager = ConfigManager::Instance();
+  ConfigManager::Print();
   //choose the Random engine
   CLHEP::HepRandom::setTheEngine(new CLHEP::RanecuEngine());
```
Do you really want to do this?  Geant4 selects a random engine itself automatically, and has a full interface for setting the engine state.

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -6,34 +6,45 @@
   G4long seed = time(NULL);
-  CLHEP::HepRandom::setTheSeed(seed);
+  //CLHEP::HepRandom::setTheSeed(seed);
+  CLHEP::HepRandom::setTheSeed(1234567890); // For debuging purposes
```
I'd get rid of this entirely.  Maybe use the Geant4 commands for initializing the random engine via macro commands?

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on app/G4_QPIX.cpp:
```diff
@@ -44,19 +55,15 @@ int main(int argc, char** argv)
-  G4RunManager* run_manager = new G4RunManager();
+  auto* run_manager = G4RunManagerFactory::CreateRunManager(G4RunManagerType::Default);
 
   G4VModularPhysicsList* physics_list = new FTFP_BERT_HP();
```
Since you used the RMFactory above, do you want to use the PhysListFactory here?

kelseymh commented on macros/Template_GENIE.mac:
```diff
@@ -0,0 +1,30 @@
+/tracking/verbose 1
+
+# configure
+/inputs/input_file /scratch/user/dave.elofson/data/GENIE_DATA/test0/test0.gtrac.root
```
Don't embed personal paths in distributed code.  I'd suggest stripping the whole path and just having it use "the current directory" when the user runs.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on macros/Template_GENIE.mac:
```diff
@@ -0,0 +1,30 @@
+/inputs/momentum_direction 0 1 0            # specifies the initial momentum of particles in detector frame
+
+# output path
+/inputs/output_file /scratch/user/dave.elofson/software/qpixg4/test0-qpixg4.root
```
Don't embed personal paths in distributed code. I'd suggest stripping the whole path and just having it use "the current directory" when the user runs.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on macros/Template_GENIE.mac:
```diff
@@ -0,0 +1,30 @@
+/inputs/output_file /scratch/user/dave.elofson/software/qpixg4/test0-qpixg4.root
+
+# initialize run
+/run/numberOfThreads 1                     # limit number of threads
```
You definitely need something, but I think it would be good to demonstrate how real "multithreaded" running works.  Maybe use four threads, which is compatible with both laptops and compute clusters?

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on macros/Template_GENIE.mac:
```diff
@@ -0,0 +1,30 @@
+/run/numberOfThreads 1                     # limit number of threads
+/run/printProgress 1
+/run/initialize
+/random/setSeeds 0 31
```
This is superseding the function calls in main().  There is no good way to do proper unique seed seeding within macro commands.  We can talk offline about maybe defining a macro command; in SuperCDMS, we defined a command like  "/random/setAutoSeed" to generate a completely unique seed for every job.

kelseymh commented on macros/Template_GENIE.mac:
```diff
@@ -0,0 +1,30 @@
+/random/setSeeds 0 31
+
+# limit radioactive decays 
+#/grdm/nucleusLimits 1 35 1 17  # aMin aMax zMin zMax
```
This command changed in G4 10.7; it's find to leave it here as an example, but you should update it to match the current G4.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on macros/single_muon.mac:
```diff
@@ -3,8 +3,11 @@
+
 # output path
-/Inputs/root_output ./output/single_muon.root
+/inputs/output_file ./output/single_muon.root
```
Is an "output/" subdirectory guaranteed to exist?  For example, does the implementation of the macro command include system calls to do "mkdir -p" if needed?

delofson0211 replied:
It is a blank directory included in the repository

kelseymh commented on macros/single_muon.mac:
```diff
@@ -18,18 +21,36 @@
 # gamma, opticalphoton, ...
 /gps/particle mu-
 
+# angular distribution
+# initial momentum in the -y direction
+#/gps/ang/type iso
+#/gps/ang/rot1 -1 0 0
+#/gps/ang/rot2 0 0 1
+#/gps/ang/mintheta 0 deg
+#/gps/ang/maxtheta 360 deg  # opening angle of cone
```
These lines are duplicated below.  Get rid of the commented-out version.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on macros/single_muon.mac:
```diff
@@ -18,18 +21,36 @@
+#/gps/pos/centre 1000 650 325 cm
+
 # angular distribution
-# initial momentum in the -y direction
```
I would put this back, and add a note that this is for the "HD configuration", so that it's clear how it relates to the subsequent commented-out block.

kelseymh commented on setup/setup_marley.sh:
```diff
@@ -18,25 +18,21 @@
-echo The initial directory is $initdir 
-marleypath=${1:-"$initdir/../marley"}
-echo The marleypath is $marleypath
+# initdir=$(pwd -P)
```
Don't leave commented out lines behind in the commit.  Just replace what's there.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on setup/setup_marley.sh:
```diff
@@ -18,25 +18,21 @@
+[ -z "$MARLEY" ] && source $1/setup_marley.sh
 
 echo Making it possible to build against marley...
 export CPLUS_INCLUDE_PATH=${CPLUS_INCLUDE_PATH}:${THIS_DIRECTORY}/include
```
Don't use "THIS_DIRECTORY" here; that's internal to the Marley setup script.  Use `${MARLEY}`.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on setup/setup_marley.sh:
```diff
@@ -18,25 +18,21 @@
 
+[ -z "$MARLEY" ] && source $1/setup_marley.sh
 
 echo Making it possible to build against marley...
```
Include the $MARLEY path in the user information (or just don't bother with any output!).

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on setup/setup_marley.sh:
```diff
@@ -18,25 +18,21 @@
 
 
-echo changing back to $initdir
-cd $initdir
-
 
```
Get rid of the extraneous blank lines at the end of the file.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/ActionInitialization.h:
```diff
@@ -0,0 +1,42 @@
+  private:
+    mutable PrimaryGeneration* masterPrimaryGeneration;
+    mutable EventAction* masterEventAction;
+    mutable RunAction* masterRunAction;
```
Get rid of this data member.  You should be getting a compiler warning about "unused variable."

delofson0211 replied:
Not getting a -Wunused_variable for this one. Leaving in for now

kelseymh commented on src/ActionInitialization.h:
```diff
@@ -0,0 +1,42 @@
+    mutable PrimaryGeneration* masterPrimaryGeneration;
+    mutable EventAction* masterEventAction;
+    mutable RunAction* masterRunAction;
+
+  private:
+    G4String   particleType;
+    G4String   string;
```
These data members are not used anywhere.  Get rid of them.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/ActionInitialization.cpp:
```diff
@@ -0,0 +1,57 @@
+// //  G4_QPIX | QPixActionInitialization.cpp
+// //
+// //
+// //   * Author: Everybody is an author!
+// //   * Creation date: 27 Oct 2022
```
Give the name of the person who actually wrote the code (and maybe an e-mail address as well).  That makes responsibility clear for who to contact if someone in the future has questions.  @delofson0211 You can look at the SuperCDMS simulation code at on Grace (~kelsey/software/supersim/) and see how we do the "top of file" comments with dated changes.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/AnalysisData.cpp:
```diff
@@ -0,0 +1,272 @@
+//-----------------------------------------------------------------------------
+AnalysisData::AnalysisData()
+{
+#ifdef G4ANALYSIS_USE
+#endif
```
Is there here because you're expecting to have additional stuff added in the future?  Maybe instead just put a comment like, "// Use a G4ANALYSIS_USE preprocessor block if necessary"?

kelseymh commented on macros/personal_macros/.GenericParticleSource_Muon_10GeV.mac.swp:
This should probably not have been added to the repository.  In fact, I'm guessing that the whole "personal_macros" diretory shouldn't be here.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/AnalysisData.cpp:
```diff
@@ -0,0 +1,272 @@
+{
+  int key = -2;
+
+  if (process.compare("NoProcess")                 == 0) key = -1;
```
Why do you use `std::string::compare()` instead of just `==`?

delofson0211 replied:
old artifact. Is the == better?

kelseymh replied:
They're equivalent in computing, but I think "==" is easier to read and understand.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions2

kelseymh commented on src/AnalysisData.cpp:
```diff
@@ -0,0 +1,272 @@
+}
+
+//-----------------------------------------------------------------------------
+int AnalysisData::ProcessToKey(std::string const & process)
```
Why are you passing the process name as a `std::string`, rather than `const G4String&`?

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/AnalysisData.h:
```diff
@@ -0,0 +1,147 @@
+    AnalysisData& operator=(const AnalysisData& rhs) = default;   
+
+
+    std::set< std::string > process_names_;
```
Is this list of names used anywhere?  I didn't see it in the .cpp file.  If it's not used, it's probably generating a compiler warning and should be removed.

delofson0211 replied:
Not producing any warnings

kelseymh commented on src/AnalysisManager.cpp:
```diff
@@ -27,320 +48,159 @@ AnalysisManager::~AnalysisManager()
 {
-    // ROOT output file
+
+  if (!G4Threading::IsMasterThread()) 
+  //G4AutoLock bookLock(&bookMutex);
+  {
+    return;
+  }  // only run Book() for the master thread
```
Do this as one line, no commented out code:
```if (!G4Threading::IsMasterThread()) return;```

Easier to read.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/AnalysisManager.cpp:
```diff
@@ -27,320 +48,159 @@ AnalysisManager::~AnalysisManager()
-    // event tree
+  // check if event_tree_ is a null pointer, if so, create event tree
+  if (event_tree_ == 0) 
+  {
```
As above, but you might even want to split this into a bunch of functions, where you pass the TTree pointer to them, and build the different pieces of the event structure.

If this was going to be a fancy ROOT-specific system, you'd probably define a class just for the event data, then use ROOT's "class dictionary" methodology, but I'm not recommending that.

kelseymh commented on src/AnalysisManager.h:
```diff
@@ -9,140 +9,56 @@
 
-        AnalysisManager();
-        ~AnalysisManager();
+    std::set< std::string > process_names_;
```
The `AnalysisManager.cpp` file is very long.  I didn't see whether this data member is used there or not.  If it is not, this data member should be removed.

delofson0211 replied:
process_names_ is used.

kelseymh commented on src/AnalysisManager.h:
```diff
@@ -9,140 +9,56 @@
+    double detector_length_y_;
+    double detector_length_z_;
+    bool detectorConfiguration_;
 
```
Delete this extraneous blank line.  Same applies in many places in the code :-)

kelseymh commented on src/ConfigManager.cpp:
```diff
@@ -0,0 +1,229 @@
+     << "Geometry -- Detector_Height:        " << detectorHeight_ << G4endl
+     << G4endl;
+}
+
+//-----------------------------------------------------------------------------
+
+
+
+
+
+
+
+
+
+
+
+
```
Get rid of this extraneous end-of-file stuff.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/ConfigManager.h:
```diff
@@ -0,0 +1,172 @@
+    G4double      detectorWidth_;
+    G4double      detectorHeight_;
+};
+
+     
+
```
Delete these extraneous blank lines.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/DetectorConstruction.cpp:
```diff
@@ -18,18 +19,19 @@
 
-
 DetectorConstruction::DetectorConstruction(): G4VUserDetectorConstruction()
-{}
```
I'm a fan of the one-line `{;}` structure, put on the same line as the declaration, but what you've changed is consistent with the rest of the code.

kelseymh commented on src/DetectorConstruction.cpp:
```diff
@@ -18,18 +19,19 @@
   // WORLD /////////////////////////////////////////////////
 
-  G4double world_size = 15.*m;
+  G4double world_size = 30.*m;
```
Should this be a configuranble parameter?  Or maybe computed from the max of the three detector dimensions?

delofson0211 replied:
probably could at some point, but I dont see this being changed at all

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/ConfigManager.h:
```diff
@@ -0,0 +1,172 @@
+    static G4double      GetEventWindow()                   { return Instance()->eventWindow_; }
+    static G4bool        GetSNTimingOn()                    { return Instance()->snTimingOn_; }
+    static G4String      GetTh2Name()                       { return Instance()->th2Name_; }
+    static G4bool        GetDetectorConfiguration()         { return Instance()->detectorConfiguration_; }
```
This needs a more understandable name if it's just returning a bool.  Maybe `UseHDDetector()`?

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/DetectorConstruction.cpp:
```diff
@@ -42,25 +44,38 @@ G4VPhysicalVolume* DetectorConstruction::Construct()
+                      
+
+  std::cout << " Detector configuration is: " << ConfigManager::GetDetectorConfiguration() << std::endl;
 
-  // DETECTOR //////////////////////////////////////////////
+  if(ConfigManager::GetDetectorConfiguration())
+  {
+  // DETECTOR HD CONFIGURATION //////////////////////////////////////////////
   // resemble an APA size
-  G4double detector_width   = 2.3*m;
-  G4double detector_height  = 6.0*m;
-  G4double detector_length  = 3.6*m;
+    ConfigManager::SetDetectorWidth(2.3*m);
+    ConfigManager::SetDetectorHeight(6.0*m);
+    ConfigManager::SetDetectorLength(3.6*m);
+  }
+  else
+  {
+  // DETECTOR VD CONFIGURATION //////////////////////////////////////////////
+    ConfigManager::SetDetectorHeight(13.0*m);
+    ConfigManager::SetDetectorLength(6.5*m);
+    ConfigManager::SetDetectorWidth(20.0*m);
+  }
```
If you move this whole block up before you create the world volume, then you can query ConfigManager to get the largest detector dimensions, and define the world volume according to that.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/EventAction.cpp:
```diff
@@ -41,81 +40,88 @@ void EventAction::BeginOfEventAction(const G4Event*)
-    if (mod == 0)
-    {
-        G4cout << "Event " << event->GetEventID() << "..." << G4endl;
+    //Replace the following lines with "/run/printProgress 1000" in the macro file
+    //int mod = g4event->GetEventID() % 1;
+    //if (mod == 0)
+    //{
+    //    G4cout << "Event " << g4event->GetEventID() << "..." << G4endl;
         // G4cout << "Energy threshold: " << energy_threshold_ << G4endl;
         // G4cout << "Total energy deposited: " << energy_deposited << G4endl;
```
Don't leave commented out code in the repository.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/GENIEManager.cpp:
```diff
@@ -0,0 +1,193 @@
+  inputFile_(""), particleType_(""), genieFormat_(""), treeName_(""),
+  isInitialized(0)
+{
+    memset(pdg_, 0, Size_*sizeof(G4int));
+    memset(vtx_, 0, Size_*sizeof(G4double));
+    memset(X4_, 0, 4*Size_*sizeof(G4double));
+    memset(P4_, 0, 4*Size_*sizeof(G4double));
+    memset(polz_, 0, 3*Size_*sizeof(G4double));
+    memset(idx, 0, Size_*sizeof(G4int));
+    memset(Status, 0, Size_*sizeof(G4int));
+    memset(rescat_, 0, Size_*sizeof(G4int));
+    memset(FirstMother, 0, Size_*sizeof(G4int));
+    memset(LastMother, 0, Size_*sizeof(G4int));
+    memset(FirstDaughter, 0, Size_*sizeof(G4int));
+    memset(LastDaughter, 0, Size_*sizeof(G4int));
```
In place of `memset()`, you can (if you want to!), use `std::fill()` using the array name; for example
```std::fill(pdg_, pdg+Size_, 0);```
The operator+() on a pointer will automatically step according to the size of the data type.

delofson0211 replied:
keeping this for now

kelseymh commented on src/GENIEManager.cpp:
```diff
@@ -0,0 +1,193 @@
+  if (genieFormat_ == "rootracker" || genieFormat_ == "grootracker") {
+    treeName_ = "gRooTracker";
+  } else {
+    G4cerr << "The genie format entered: '" << genieFormat_ << "' is not currently supported." << G4endl
+           << "Please make the necessary adjustments in GENIEManager.cpp and GENIEManager.h and try again." << G4endl
+           << G4endl
+           << "Current list of supported genie formats:" << G4endl
+           << " - rootracker" << G4endl;
+    abort();
```
Since this is a Geant4 program, you might consider using G4Exception with the "FatalException" enum value.  Not required.

kelseymh commented on src/GENIEManager.cpp:
```diff
@@ -0,0 +1,193 @@
+  if (!inputFile_.empty())
+  {
+    isInitialized=0;
+    G4cout << "Configuring to Read From ROOT File..." << G4endl;
```
Do you want to protect diagnostic output with an "if (verbose)" test?  You could have a verbosity parameter in `ConfigManager` to be used globally in the code, or you could have each "top-level" object configurable with it's own verbosity parameter.

delofson0211 replied:
Want to add a verbosity parameter in next upgrade

kelseymh commented on src/GENIEManager.cpp:
```diff
@@ -0,0 +1,193 @@
+      if (!f_ || !f_->IsOpen()) {
+        f_ = new TFile(inputFile_);
+      }
+      if(f_->IsZombie()){
```
Do you want to use `IsZombie()` or `!IsGood()`?  I'm not sure Zombie covers all of the failure modes, but it might.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/GENIEManager.cpp:
```diff
@@ -0,0 +1,193 @@
+
+
+void GENIEManager::Cd(){
+  if(f_)
+    f_->cd();
```
Two comments:  Why not use a one-line statement here?  Insert a space between "if" and "(".

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/GENIEManager.h:
```diff
@@ -0,0 +1,136 @@
+    G4String        genieFormat_;
+    G4String        treeName_;
+    G4int           isInitialized;
+
+
+
+
+
```
Strip out these extraneous spaces.  There's a lot of them throughout :-)

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/PrimaryGeneration.cpp:
```diff
@@ -186,9 +200,188 @@ void PrimaryGeneration::GeneratePrimaries(G4Event* event)
+      G4Exception("[PrimaryGeneration]","GENIEGeneratePrimaries",G4ExceptionSeverity::JustWarning,str1.c_str());
+      continue;
+    }
+
+    // Print out info about the GENIE particle
+    //G4cout << "..........oooooOOO000OOOooooo.........." << G4endl
+    //       << "GENIE Particle Info:" << G4endl
+    //       << "  PDG Code:       " << genieManager->GetPDG_(np) << G4endl
+    //       << "  Energy:         " << genieManager->GetE_(np) << G4endl
+    //       << "  Px:             " << genieManager->GetPx_(np) << G4endl
+    //       << "  Py:             " << genieManager->GetPy_(np) << G4endl
+    //       << "  Pz:             " << genieManager->GetPz_(np) << G4endl
+    //       << "G4ParticleDefinition Info:" << G4endl
+    //       << "  Mass:           " << pdef->GetPDGMass() << G4endl
+    //       << "  Charge:         " << pdef->GetPDGCharge() << G4endl
+    //       << "User defined Info:" << G4endl
+    //       << "  Vertex:         " << vertex3d << G4endl
+    //       << "..........oooooOOO000OOOooooo.........." << G4endl;
+
+
+    // filter the particles less than 1 eV
+    // if(genieManager->GetE_(np)<(1 *CLHEP::eV)) continue;
```
Don't keep commented out code in the repository.  You could protect the printout with a verbosity flag.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/PrimaryGeneration.cpp:
```diff
@@ -463,3 +656,51 @@ void PrimaryGeneration::MARLEYGeneratePrimaries(G4Event* event)
 
 }
+
+ROOT::Math::SMatrix< double, 3 > PrimaryGeneration::Rotation_Matrix(G4ThreeVector vec1, G4ThreeVector vec2)
```
You should check the `CLHEP/Vector/RotationMatrix.h` file and CLHEP documentation.  They are likely to already have a function that does this computation for you, so you could use a simple `G4RotationMatrix` object, rather than writing your own.

kelseymh commented on src/SteppingAction.cpp:
```diff
@@ -27,5 +27,10 @@ void SteppingAction::UserSteppingAction(const G4Step* step)
     AnalysisManager * analysis_manager = AnalysisManager::Instance();
 
-    analysis_manager->AddProcess(step->GetPostStepPoint()->GetProcessDefinedStep()->GetProcessName());
+    if (step->GetPostStepPoint()->GetProcessDefinedStep() != 0){
+      event.AddProcess(step->GetPostStepPoint()->GetProcessDefinedStep()->GetProcessName());
+    } else {
+      event.AddProcess("User Limit");
+    }
+
```
Why not get the process pointer into a local variable, reducing the copy-pasted code:
```
G4VProcess* stepProc = step->GetPostStepPoint()->GetProcessDefinedStep();
G4String procName = (stepProc ? stepProc->GetProcessName() : "User Limit");
analysis_manager->AddProcess(procName);
```

kelseymh commented on src/TrackingAction.cpp:
```diff
@@ -59,6 +59,12 @@ void TrackingAction::PreUserTrackingAction(const G4Track* track)
         )
     );
 
+    //G4cout << "Track info: " << G4endl
+    //       << "  Track ID:       " << track->GetTrackID() << G4endl
+    //       << "  Track PDG:      " << track->GetDefinition()->GetPDGEncoding() << G4endl
+    //       << "  Track Momentum: " << track->GetMomentum().x() << ", " << track->GetMomentum().y() << ", " << track->GetMomentum().z() << G4endl
+    //       << "  Track Energy:   " << track->GetTotalEnergy() << G4endl; 
+
```
Don't leave commented-out code in the repository.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

kelseymh commented on src/TrackingAction.cpp:
```diff
@@ -82,6 +89,10 @@ void TrackingAction::PostUserTrackingAction(const G4Track* track)
 
     // set process
-    particle->SetProcess(track->GetStep()->GetPostStepPoint()->GetProcessDefinedStep()->GetProcessName());
+    if (track->GetStep()->GetPostStepPoint()->GetProcessDefinedStep() != 0) {
+      particle->SetProcess(track->GetStep()->GetPostStepPoint()->GetProcessDefinedStep()->GetProcessName());
+    } else {
+      particle->SetProcess("User Limit");
+    }
 }
```
See my comment above on the same piece of code.

kelseymh reviewed (changes requested):
A general comment for the future:  It is generally better to have the code review done at the point where you are merging the feature branch onto develop, rather than merging develop onto master.  

As it is now, any corrections you make as part of this review should _formally_ be done on a new feature branch, then merged back onto develop and evaluated again.

Committing code changes directly to develop is generally discouraged.

asanchezcastillo commented on src/RunAction.cpp:
```diff
@@ -71,28 +111,34 @@ void RunAction::BeginOfRunAction(const G4Run* run)
-        std::string parent_path = path.parent_path();
-        std::string stem = path.stem();
-        std::string extension = path.extension();
+        G4String parentPath_ = outputFile_(0, outputFile_.last('/'));
```
I am having issues to compile the code with my Geant4 version (11.1.0) as the member `last` in `G4String` seems to have been removed from version 4.11 onwards.

kelseymh replied:
@asanchezcastillo Is qpixg4 supposed to be working with Geant4 11 at all?  My own experience (with the SuperCDMS simulation), is that if you migrate code to compile and run with G4 11, then it will not compile with G4 10.7, and vice versa.

I thought Q-Pix had decided (at least for a while) to stick with G4 10.7, to avoid those problems.

asanchezcastillo replied:
@kelseymh there is a google doc containing an installation guide for all the qpix software in which G411.1.0 is used. I also found out in the qpixg4 wiki one of the dependencies for qpixg4-v2.0.0 is Geant4 11.0+. I particularly do not mind using G410.7 but I do not know if it could be inconvenient for other people.

kelseymh replied:
Ah, thank you!  Then I was mistaken.  I do see that `.last()` was removed from G4String with G4 11.  @delofson0211 You may want to look at replacing `.last()` with a use of `.rbegin()` or something directly from `std::string` (which is the base class of `G4String`).

delofson0211 replied:
@asanchezcastillo qpixg4 is made specifically to be compatible with Geant4.10.7, and not with Geant4.11. There are many breaking changes. The wiki needs to be changed (I think we were attempting to make it compatible and changed it before we realized it was going to be a pain). Making it compatible with Geant4.11 will be a project for later (and someone else)

kelseymh replied:
@delofson0211 If this branch of qpixg4 does build and run against G4 11, this sounds like a small fix to keep it that way.  If it doesn't, then what you say above is exactly right.

delofson0211 replied:
Yes. There are no branches that do run against Geant4.11. It was supposed to be an easy fix, but it turned out to be much more complicated.

delofson0211 committed:
changed the quotations and brackets in the includes of app/G4_QPIX.cpp, add label comments to the includes

delofson0211 committed:
remove superfluous blank space, remove print statement from before variables are set

delofson0211 committed:
no longer assigning ConfigManager::Instance to variable to remove unused variable warning

kelseymh commented on Build/.keep:
Removing this file means that Git will no longer create a "Build" directory when the user creates a clone.  I personally think that's the right approach.  The user should create their CMake building and installing directories wherever is appropriate for them, and CMake building should _not_ be done within the Git-controlled source directory.  But this is a change of behaviour, and you should be sure that the build scripts still work properly if "qpixg4/Build" is not present.

delofson0211 committed:
removing personal paths from macro file

delofson0211 committed:
removing personal paths from macro file

delofson0211 committed:
removing duplicate commented-out angular distribution lines

delofson0211 committed:
removing commented out lines

delofson0211 committed:
changing THIS_DIRECTORY to MARLEY

delofson0211 committed:
removed blank spaces at eof, output the marley path

delofson0211 committed:
removing unused variables

delofson0211 committed:
change FillMetadata from taking arguments to pulling from ConfigManager, added comments to DetectorConstruction to accompany detector dimensions

delofson0211 committed:
change /grdm/ to updated command /process/had/rdm/

delofson0211 committed:
Add author info

delofson0211 committed:
stopped tracking macros/personal_macros directory

delofson0211 committed:
changing order of data members to remove -Wreorder warnings

delofson0211 committed:
change data member type of ProcessToKey argument from std::string & to const G4String &

delofson0211 committed:
fixed the -Wshadow of event_id_offset

delofson0211 committed:
fixed the -Wshadow of energy_threshold

delofson0211 committed:
fix typo

delofson0211 committed:
fix abort calls

delofson0211 commented on src/PrimaryGeneration.cpp:
It appears #include "G4SystemOfUnits.h" is causing lots of -Wshadow warnings regarding the units between ROOT and CLHEP. Need to fix these

kelseymh replied:
Yup.  That's a know collision that we have run into in SuperCDMS.  The underlying cause is that G4SystemOfUnits.h is nothing but a whole bunch of `using` directives, which strip off the CLHEP:: namespace from unit multipliers (e.g., `using CLHEP::MeV;` means that in your code you can just write `0.511*MeV` instead of `0.511*CLHEP::MeV`, and so on).

There are things you can do to help reduce the shadow warning:

1) Don't "ever" put G4SystemOfUnits.h into other header files.  If you really need a unit multipler in a .h file (for example, to set a default value for a function argument), then use `#include "CLHEP/Units/SystemOfUnits.h"`, and put the "CLHEP::" namespace where you need it.

2) In your .cpp files, avoid having G4SystemOfUnits.h at the same time as any ROOT classes.  The shadow warnings come from very low level ROOT .h files, so you can't avoid them with your ROOT code.  If you are not actually _using_ unit multipliers, then you don't need the .h file!

3) If you do need unit multipliers in a .cpp file that includes ROOT, then follow (1) above and switch to the CLHEP namespace and header file.

kelseymh replied:
Weird.  GitHub won't let me add a comment to line 26 below; it's greyed out.  Remove line 26.  You may replace it with `#include "CLHEP/Units/SystemOfUnits.h"` as above.  Unless I missed something, someone in the past already went through PrimaryGeneration.cpp and added the CLHEP:: namespace where needed.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

delofson0211 committed:
removing commented out lines

delofson0211 committed:
remove unneeded messenger from .h

kelseymh commented on src/PrimaryGeneration.cpp:
```diff
@@ -47,25 +50,16 @@
-    vertex_y_(6.0/2),
-    vertex_z_(3.6/2),
-    particle_gun_(0)
+    particle_gun_(0)//,
+    //isotropic_(false),
+    //decay_at_time_zero_(false),
+    //override_vertex_position_(false),
+    //vertex_x_(0),
+    //vertex_y_(0),
+    //vertex_z_(0),
+    //printParticleInfo_(false),
+    //particleType_("")
```
Don't include commented out code in the repository.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions branch

delofson0211 committed:
removing commented out code

delofson0211 committed:
change nThreads to 4 to show how to use different nThreads

delofson0211 committed:
changing G4SystemOfUnits to CLHEP/Units/SystemOfUnits and changing necessary unit strings

delofson0211 committed:
deal with -Wshadow extra variables

delofson0211 committed:
remove unused variables

delofson0211 committed:
remove unused variables and shadows

delofson0211 committed:
switching the G4SystemOfUnits to CLHEP/Units/SystemOfUnits

delofson0211 committed:
fixed some typos

delofson0211 committed:
added forward declaration to .h, added the CLHEP:: prefix to more unit strings

delofson0211 committed:
removing unused variables, moving Instance() calls to beginning

delofson0211 committed:
remove artifact detector dims metadata code, already replaced with ConfigManager

delofson0211 committed:
formatting

delofson0211 committed:
removing extraneous blank space

delofson0211 committed:
removing extraneous blank space

delofson0211 committed:
Change detectorConfiguration variable name and messenger names to be more descriptive

delofson0211 committed:
allow for varying world size to correspond to size of chosen detector

delofson0211 committed:
removing commented out lines

delofson0211 committed:
reformatting and removing blank spaces

delofson0211 committed:
removing commented out lines

delofson0211 committed:
Changing IsZombie() to notIsGood() to be more robust

delofson0211 committed:
Merge pull request #20 from Q-Pix/bugfix/v3.0.0_revisions

Bugfix/v3.0.0 revisions

kelseymh commented on src/ConfigManager.cpp:
```diff
@@ -0,0 +1,212 @@
+#include "G4GenericMessenger.hh"
+#include "G4Threading.hh"
+#include "G4ThreeVector.hh"
+
```
Delete extra blank lines?

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions2

kelseymh commented on src/ConfigManager.cpp:
```diff
@@ -0,0 +1,212 @@
+  particleType_(""), decayAtTimeZero_(false), isotropic_(true),
+  overrideVertexPosition_(false), printParticleInfo_(false), inputFile_(""), outputFile_(""), marleyJson_(""), generator_(""),
+  genieFormat_(""), multirun_(false), momentumDirection_(0,0,0), vertexX_(2.3*CLHEP::m/2), vertexY_(6.0*CLHEP::m/2), vertexZ_(3.7*CLHEP::m/2),
+  
+  nAr39Decays_(0), nAr42Decays_(0), nKr85Decays_(0), nCo60Decays_(0), nK40Decays_(0),
+  nK42Decays_(0), nBi214Decays_(0), nPb214Decays_(0), nPo210Decays_(0), nRn222Decays_(0), eventCutoff_(0),
+  eventWindow_(0),
+
+  snTimingOn_(false), th2Name_("nusperbin2d_nue"),
+
```
Delete the extra blank lines.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions2

kelseymh commented on src/ConfigManager.cpp:
```diff
@@ -0,0 +1,212 @@
+//-----------------------------------------------------------------------------
+ConfigManager::ConfigManager(const ConfigManager& master)
+  : eventIDOffset_(master.eventIDOffset_),
+  energyThreshold_(master.energyThreshold_),
+
+  particleType_(master.particleType_),
+  decayAtTimeZero_(master.decayAtTimeZero_), isotropic_(master.isotropic_),
+  overrideVertexPosition_(master.overrideVertexPosition_),
+  printParticleInfo_(master.printParticleInfo_), inputFile_(master.inputFile_),
+  outputFile_(master.outputFile_), marleyJson_(master.marleyJson_),
+  generator_(master.generator_), genieFormat_(master.genieFormat_),
+  multirun_(master.multirun_), momentumDirection_(master.momentumDirection_),
+  vertexX_(master.vertexX_), vertexY_(master.vertexY_),
+  vertexZ_(master.vertexZ_),
+
+  nAr39Decays_(master.nAr39Decays_), nAr42Decays_(master.nAr42Decays_),
+  nKr85Decays_(master.nKr85Decays_), nCo60Decays_(master.nCo60Decays_),
+  nK40Decays_(master.nK40Decays_), nK42Decays_(master.nK42Decays_),
+  nBi214Decays_(master.nBi214Decays_), nPb214Decays_(master.nPb214Decays_),
+  nPo210Decays_(master.nPo210Decays_), nRn222Decays_(master.nRn222Decays_),
+  eventCutoff_(master.eventCutoff_), eventWindow_(master.eventWindow_),
+
+  snTimingOn_(master.snTimingOn_), th2Name_(master.th2Name_),
+
+  useHDDetectorConfiguration_(master.useHDDetectorConfiguration_), detectorLength_(master.detectorLength_),
+  detectorWidth_(master.detectorWidth_), detectorHeight_(master.detectorHeight_)
```
Delete extra blank lines.

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions2

kelseymh commented on src/DetectorConstruction.h:
```diff
@@ -19,9 +19,11 @@ class DetectorConstruction: public G4VUserDetectorConstruction
 private:
   virtual G4VPhysicalVolume* Construct();
   virtual void ConstructSDandField();
+
```
Another extra blank line...

delofson0211 replied:
fixing in bugfix/v3.0.0_revisions2

delofson0211 committed:
changing std::string.compare to ==

delofson0211 committed:
removed extra blank lines

delofson0211 committed:
Merge pull request #21 from Q-Pix/bugfix/v3.0.0_revisions2

Bugfix/v3.0.0 revisions2

delofson0211 committed:
fixing typo

delofson0211 committed:
fixing IsGood->IsOpen
